django/db/migrations/executor.py

```diff
@@ -87,7 +87,7 @@ def migrate(self, targets, plan=None, fake=False, fake_initial=False):
         for migration, _ in full_plan:
             if migration in migrations_to_run:
                 states[migration] = state.clone()
-            state = migration.mutate_state(state)  # state is cloned inside
+            state = migration.mutate_state(state, preserve=False)
```

Comment:
Double checking the commit, this change, in this form leaks some state across migrations. Testing on CI right now.

```python
ERROR: test_alter_id_type_with_fk (migrations.test_executor.ExecutorTests)
----------------------------------------------------------------------
Traceback (most recent call last):
  File "/home/markus/Coding/django/django/db/backends/utils.py", line 64, in execute
    return self.cursor.execute(sql, params)
psycopg2.ProgrammingError: foreign key constraint "book_app_book_author_id_93532c48_fk_author_app_author_id" cannot be implemented
DETAIL:  Key columns "author_id" and "id" are of incompatible types: integer and character varying.


The above exception was the direct cause of the following exception:

Traceback (most recent call last):
  File "/home/markus/Coding/django/django/test/utils.py", line 182, in inner
    return test_func(*args, **kwargs)
  File "/home/markus/Coding/django/tests/migrations/test_executor.py", line 401, in test_alter_id_type_with_fk
    executor.migrate([("author_app", "0002_alter_id")])
  File "/home/markus/Coding/django/django/db/migrations/executor.py", line 94, in migrate
    self.apply_migration(states[migration], migration, fake=fake, fake_initial=fake_initial)
  File "/home/markus/Coding/django/django/db/migrations/executor.py", line 131, in apply_migration
    state = migration.apply(state, schema_editor)
  File "/home/markus/Coding/django/django/db/migrations/migration.py", line 118, in apply
    operation.database_forwards(self.app_label, schema_editor, old_state, project_state)
  File "/home/markus/Coding/django/django/db/migrations/operations/fields.py", line 201, in database_forwards
    schema_editor.alter_field(from_model, from_field, to_field)
  File "/home/markus/Coding/django/django/db/backends/base/schema.py", line 482, in alter_field
    old_db_params, new_db_params, strict)
  File "/home/markus/Coding/django/django/db/backends/base/schema.py", line 635, in _alter_field
    params,
  File "/home/markus/Coding/django/django/db/backends/base/schema.py", line 106, in execute
    cursor.execute(sql, params)
  File "/home/markus/Coding/django/django/db/backends/utils.py", line 79, in execute
    return super(CursorDebugWrapper, self).execute(sql, params)
  File "/home/markus/Coding/django/django/db/backends/utils.py", line 64, in execute
    return self.cursor.execute(sql, params)
  File "/home/markus/Coding/django/django/db/utils.py", line 95, in __exit__
    six.reraise(dj_exc_type, dj_exc_value, traceback)
  File "/home/markus/Coding/django/django/utils/six.py", line 658, in reraise
    raise value.with_traceback(tb)
  File "/home/markus/Coding/django/django/db/backends/utils.py", line 64, in execute
    return self.cursor.execute(sql, params)
django.db.utils.ProgrammingError: foreign key constraint "book_app_book_author_id_93532c48_fk_author_app_author_id" cannot be implemented
DETAIL:  Key columns "author_id" and "id" are of incompatible types: integer and character varying.
```

However, integrating this with the second commit on my PR fixes the issue. I thus squash those commits there and close your PR here.